feat(global): initialize Django project with session management and e…

abolfazljazak opened the pull request:
…ncryption modules

- Add `.gitignore` to exclude unnecessary files from version control.
- Initialize the `pyproject.toml` and `poetry.lock` files for dependency management using Poetry.
- Add `requirements.txt` for compatibility with pip-based environments.
- Create the `sage_session` app with a modular structure.
- Implement base encryption classes and a `FernetEncryptor` using the `cryptography` package.
- Add `SessionHandler` class to manage session variables with encryption and custom expiration times.

abolfazljazak committed:
feat(global): initialize Django project with session management and encryption modules

- Add `.gitignore` to exclude unnecessary files from version control.
- Initialize the `pyproject.toml` and `poetry.lock` files for dependency management using Poetry.
- Add `requirements.txt` for compatibility with pip-based environments.
- Create the `sage_session` app with a modular structure.
- Implement base encryption classes and a `FernetEncryptor` using the `cryptography` package.
- Add `SessionHandler` class to manage session variables with encryption and custom expiration times.

sepehr-akbarzadeh commented on pyproject.toml:
```diff
@@ -0,0 +1,17 @@
+
+[tool.poetry.dependencies]
+python = "^3.11"
+django = "4.2"
```
4.2 and above. It must also support 5 and new versions

sepehr-akbarzadeh commented on sage_session/apps.py:
```diff
@@ -0,0 +1,6 @@
+
+class SageSessionConfig(AppConfig):
+    default_auto_field = "django.db.models.BigAutoField"
+    name = "sage_session"
```
Add verbosr name with translation after this

sepehr-akbarzadeh commented on sage_session/encryptors/base.py:
```diff
@@ -0,0 +1,50 @@
+            The encrypted data.
+
+        """
+        pass
```
Instead of pass use

```py
raise notImplementrdError()
```

sepehr-akbarzadeh commented on sage_session/encryptors/fernet_encrypt.py:
```diff
@@ -0,0 +1,112 @@
+try:
+    from cryptography.fernet import Fernet
+except ImportError:
+    raise ImportError("Install `cryptography` package. Run `pip install cryptography`.")
```
Improve message here.

For example: to use `pkg name` encryption feature enable add cryptography ...

sepehr-akbarzadeh commented on sage_session/handlers/session.py:
```diff
@@ -0,0 +1,122 @@
+import logging
```
Add unit test for services

abolfazljazak committed:
fix(pyproject.toml): configure Commitizen with conventional commits and Semantic Versioning

- Added [tool.commitizen] section in `pyproject.toml` to set up Commitizen using conventional commits.
- Configured `tag_format` as "$version" to standardize version tags.
- Set `version_scheme` to "semver" for Semantic Versioning.
- Integrated Poetry as the `version_provider` to align versioning with project dependencies.
- Enabled `update_changelog_on_bump` to automatically update the changelog with each version bump.
- Configured `major_version_zero` to true to allow breaking changes during 0.x versions.

abolfazljazak committed:
fix(pyproject.toml): update Django version constraint in pyproject.toml

Updated the Django version constraint in `pyproject.toml` to `>=4.2,<6.0` to ensure compatibility with the latest Django 4.x versions while avoiding major version 6.x.

abolfazljazak committed:
feat(apps): add verbose_name to Sage Session Management app

Added the `verbose_name` attribute to the `Sage Session Management` app in `sage_session/apps.py` to provide a human-readable name for the application in Django admin and other interfaces.

abolfazljazak committed:
refactor(encryptors): replace pass with NotImplementedError in Encryptor class

Replaced `pass` statements with `raise NotImplementedError` in the `encrypt` and `decrypt` methods of the `Encryptor` class in `sage_session/encryptors/base.py`. This change ensures that subclasses of `Encryptor` will explicitly implement these methods, enforcing proper implementation.

abolfazljazak committed:
refactor(encryptors): improve ImportError message for cryptography package

Refined the `ImportError` message in `sage_session/encryptors/fernet_encrypt.py` to provide clearer instructions for installing the `cryptography` package. The revised message offers more detailed guidance for resolving the missing package issue.

abolfazljazak committed:
build(global): add Tox configuration and update dependencies

- Added a new `tox.ini` file to configure testing environments for multiple Python and Django versions.
- Updated `requirements.txt` and `pyproject.toml` to include `tox` for testing management.
- Modified `poetry.lock` to reflect the addition of `tox`.

abolfazljazak committed:
build(pytest): add pytest configuration and update dependencies

- Added a new `pytest.ini` file to configure pytest for Django settings.
- Updated `pyproject.toml` and `poetry.lock` to include `pytest` and `pytest-django` packages.

abolfazljazak committed:
test(sage_session): add unit tests for SessionHandler

- Added unit tests in `sage_session/tests/handlers/test_session.py` to cover the following scenarios:
  - Storing and retrieving a session value.
  - Storing a session value with an expiry.
  - Deleting a session value.
  - Checking the existence of a session value.
  - Storing and retrieving a session value without encryption.
- Included necessary fixtures and setup in `sage_session/tests/__init__.py` and `sage_session/tests/handlers/__init__.py`.

abolfazljazak committed:
build(poetry): add code formatting and linting tools

- Updated `pyproject.toml` and `poetry.lock` to include `isort`, `black`, `pre-commit`, and `ruff`.
- These tools are added to standardize code formatting and improve code quality across the project.

abolfazljazak committed:
build(poetry): add mypy and pylint for static code analysis

- Updated `pyproject.toml` and `poetry.lock` to include `mypy` and `pylint`.
- These tools are added to perform static type checking and linting for improving code quality and maintaining consistency.

abolfazljazak committed:
build(pre-commit): add pre-commit hooks configuration

- Added a `.pre-commit-config.yaml` file to set up pre-commit hooks for various tasks, including:
  - Code formatting with `black`, `isort`, and `docformatter`.
  - Linting with `ruff`, `pylint`, and `bandit`.
  - Managing project files with `pyupgrade`, `pyproject-fmt`, and `tox-ini-fmt`.
  - Documentation checks with `blacken-docs` and `rstcheck`.
  - Commit message enforcement with `commitizen`.
  - Added custom hooks for running `pytest` and `pylint` during commit.
- Updated `pyproject.toml` to configure `pylint` and other tools.

abolfazljazak committed:
feat(sage_session): add translation support for app verbose_name

- Updated `SageSessionConfig` to support translation for `verbose_name` using Django's `gettext_lazy`.

abolfazljazak committed:
refactor(sage_session): include DummyEncryptor in module exports

- Updated `sage_session/encryptors/__init__.py` to include `DummyEncryptor` in the module's `__all__` exports.
